### ISSUE TITLE
It has been a long journey

### ISSUE THREAD

**Issue by eznix86** [Emmanuel-cangy/git-it#5]
### REST MY FRIEND, REST

![gandalf](https://c.tenor.com/ff_3wbQAtLEAAAAd/gandalf-sly.gif)

- [x] you are done !

TLDR; 
- you learn about git push, merge, delete after merge, pull request, resolve a conflict and rebase

ANOTHER TDLR;
- When you have a small commit branch, like few lines, you can merge main onto your branch and submit a pull request
- Whenever you have big branch wit lots of commits (10+ files) (which is not recommended), it is better to rebase.
- A rebase happens mostly when you changed a file which no longer is the same when you created the branch. So, git doesn't know if is it is okay to merge your changes when the file is not the same as before. Conflicts happens. A lot. When your branch is older than the current state of the branch you are merging onto.
- So, keep your commit branch small, or simple.
   - take your task
   - split it into small branch
   - and merge it little by little into a big feature branch which sync from time to time on the main branch

```mermaid
    gitGraph
       checkout main
       commit
       commit
       branch feature_branch
       commit
       commit
       checkout main
       commit
       commit
       checkout feature_branch
       branch feature_task_1
       checkout feature_task_1
       commit
       commit
       checkout feature_branch
       merge main
       merge feature_task_1
       branch feature_task_2
       checkout feature_task_2
       commit
       commit
       checkout feature_branch
       merge main
       merge feature_task_2
       branch feature_task_3
       checkout feature_task_3
       commit
       commit
       checkout feature_branch
       merge main
       merge feature_task_3
       checkout feature_branch
       commit
       commit
       checkout main
       merge feature_branch
  ```
 - In the image above, we have `feature_branch` and `main`
 - when you are doing your feature split them into tasks, create multiple branch
 - here we create task_1 we do our stuff, then before we merge we sync feature_branch with main, then we merge task_1 onto the feature_branch, then we create our next task, and sync feature_branch and merge the next task and so on.
 - In the end, we will have a big branch called feature_branch with small branches inside already merged, and since, feature_branch isn't that old, because every time we start a task, we sync it, you can just merge it into main without problem.
 - Also, there are organisation which omit a feature_branch, they just start the small branch from main and merge it back directly, this is a Trunk-based development.  [^1]

[^1]: There are multiple strategies for git: https://launchdarkly.com/blog/git-branching-strategies-vs-trunk-based-development/.

**Comment by Emmanuel-cangy** [comment]
Was nice
